### PR TITLE
Fix `test_token_networks`

### LINKED ISSUE #1796: operator '<' not supported for 'TransactionChannelNewBalance'
This  error  terminates my tests after latest rebase:

```
Traceback (most recent call last):
  File "src/gevent/greenlet.py", line 716, in gevent._greenlet.Greenlet.run
  File "/home/xoza/src/raiden/raiden/tasks.py", line 56, in _run
    self.poll_for_new_block()
  File "/home/xoza/src/raiden/raiden/tasks.py", line 95, in poll_for_new_block
    result = callback(current_block, chain_id)
  File "/home/xoza/src/raiden/raiden/raiden_service.py", line 376, in _callback_new_block
    on_blockchain_event(self, event, current_block_number, chain_id)
  File "/home/xoza/src/raiden/raiden/blockchain_events_handler.py", line 250, in on_blockchain_event
    handle_channel_new_balance(raiden, event, current_block_number)
  File "/home/xoza/src/raiden/raiden/blockchain_events_handler.py", line 153, in handle_channel_new_balance
    raiden.handle_state_change(newbalance_statechange, current_block_number)
  File "/home/xoza/src/raiden/raiden/raiden_service.py", line 332, in handle_state_change
    event_list = self.wal.log_and_dispatch(state_change, block_number)
  File "/home/xoza/src/raiden/raiden/storage/wal.py", line 47, in log_and_dispatch
    events = self.state_manager.dispatch(state_change)
  File "/home/xoza/src/raiden/raiden/transfer/architecture.py", line 133, in dispatch
    state_change,
  File "/home/xoza/src/raiden/raiden/transfer/node.py", line 721, in state_transition
    state_change,
  File "/home/xoza/src/raiden/raiden/transfer/node.py", line 412, in handle_token_network_action
    chain_state.block_number,
  File "/home/xoza/src/raiden/raiden/transfer/token_network.py", line 228, in state_transition
    block_number,
  File "/home/xoza/src/raiden/raiden/transfer/token_network.py", line 87, in handle_balance
    block_number,
  File "/home/xoza/src/raiden/raiden/transfer/token_network.py", line 32, in subdispatch_to_channel_by_id
    block_number,
  File "/home/xoza/src/raiden/raiden/transfer/channel.py", line 1845, in state_transition
    block_number,
  File "/home/xoza/src/raiden/raiden/transfer/channel.py", line 1757, in handle_channel_newbalance
    heapq.heappush(channel_state.deposit_transaction_queue, order)
TypeError: '<' not supported between instances of 'TransactionChannelNewBalance' and 'TransactionChannelNewBalance'
2018-07-12T14:05:12Z <AlarmTask "Greenlet-0" at 0x7f0ca0842848: _run> failed with TypeError
```

it happens because heap can't order a namedtuple that contains `TransactionChannelNewBalance`
```
order = TransactionOrder(
            deposit_transaction.deposit_block_number,
            deposit_transaction,
        )
heapq.heappush(channel_state.deposit_transaction_queue, order) # fail
```

### LINKED ISSUE #1819: Exception on /api/1/connections/0xaddress [PUT]
## Problem Definition

When trying to join a token network from the WebUI

```
2018-07-13 15:22:48 [error    ] Exception on /api/1/connections/0x2F4Aaa4112E1891cC32b630e167a7179cFbE7b03 [PUT] [flask.app]
Traceback (most recent call last):
  File "/Users/paul/Work/venv-raiden/lib/python3.6/site-packages/flask/app.py", line 1813, in full_dispatch_request
    rv = self.dispatch_request()
  File "/Users/paul/Work/venv-raiden/lib/python3.6/site-packages/flask/app.py", line 1799, in dispatch_request
    return self.view_functions[rule.endpoint](**req.view_args)
  File "/Users/paul/Work/venv-raiden/lib/python3.6/site-packages/flask_restful/__init__.py", line 480, in wrapper
    resp = resource(*args, **kwargs)
  File "/Users/paul/Work/venv-raiden/lib/python3.6/site-packages/flask/views.py", line 88, in view
    return self.dispatch_request(*args, **kwargs)
  File "/Users/paul/Work/venv-raiden/lib/python3.6/site-packages/flask_restful/__init__.py", line 595, in dispatch_request
    resp = meth(*args, **kwargs)
  File "/Users/paul/Work/venv-raiden/lib/python3.6/site-packages/webargs/core.py", line 441, in wrapper
    return func(*args, **kwargs)
  File "/Users/paul/Work/raiden/raiden/api/v1/resources.py", line 186, in put
    joinable_funds_target=joinable_funds_target,
  File "/Users/paul/Work/raiden/raiden/api/rest.py", line 462, in connect
    joinable_funds_target,
  File "/Users/paul/Work/raiden/raiden/api/python.py", line 164, in token_network_connect
    joinable_funds_target=joinable_funds_target,
  File "/Users/paul/Work/raiden/raiden/connection_manager.py", line 132, in connect
    self._open_channels()
  File "/Users/paul/Work/raiden/raiden/connection_manager.py", line 296, in _open_channels
    partner,
TypeError: channel_open() missing 1 required positional argument: 'partner_address'
```

### PR DESCRIPTION
**Pull request by pcppcp**
- fixes a race condition when opening a new channel
- updates some outdated api used in test and connection manager

~~- there are more races (the test is flaky) waiting to be fixed - I am working on that now.~~
- fixes #1796 
- fixes #1819

### CODE REVIEW

File: raiden/connection_manager.py [hunk 1]
```diff
@@ -292,17 +294,29 @@ def _open_channels(self):
...
+                    self.registry_address,
+                    self.token_address,
                     partner,
                 )
             except DuplicatedChannelError:
                 # This can fail because of a race condition, where the channel
                 # partner opens first.
                 log.info('partner opened channel first')
 
+            exception = BaseException('Internal state did not sync an existing channel')
```
Comment (hackaugusto) [inline raiden/connection_manager.py]: What about `RuntimeError`?

File: raiden/connection_manager.py [hunk 2]
```diff
@@ -292,17 +294,29 @@ def _open_channels(self):
...
+            with gevent.Timeout(self.channel_sync_state_timeout, exception=exception):
+                waiting.wait_for_newchannel(
```
Comment (hackaugusto) [inline raiden/connection_manager.py]: [`RaidenApi.channel_open`](https://github.com/raiden-network/raiden/blob/c72407fd6c7d1f6e01dc703a1ba6278091038ca2/raiden/api/python.py#L231-L238) is already calling `wait_for_newchannel`. This is not be necessary

File: raiden/tests/integration/long_running/flaky/test_token_networks.py [hunk 3]
```diff
@@ -51,57 +92,70 @@ def test_participant_selection(raiden_network, token_addresses):
...
-        ]
-
-    assert all(open_channels_count(connection_managers))
-
-    def not_saturated(connection_managers_):
-        return [
-            1 for connection_manager_ in connection_managers_
-            if len(connection_manager_.open_channels) < connection_manager_.initial_channel_target
-        ]
+    assert all([x() for x in open_channel_views])
```
Comment (hackaugusto) [inline raiden/tests/integration/long_running/flaky/test_token_networks.py]: the inner list is not required

File: raiden/tests/integration/long_running/flaky/test_token_networks.py [hunk 4]
```diff
@@ -51,57 +92,70 @@ def test_participant_selection(raiden_network, token_addresses):
...
-    for _ in range(wait_blocks):
-        for app in raiden_network:
-            wait_until_block(
-                app.raiden.chain,
-                app.raiden.chain.block_number() + 1,
-            )
+    for app in raiden_network:
+        wait_until_block(
+            app.raiden.chain,
+            app.raiden.chain.block_number() + 1,
```
Comment (hackaugusto) [inline raiden/tests/integration/long_running/flaky/test_token_networks.py]: Why decrease the number of waiting blocks? p.s.: I don't know why it was 15 blocks previously.

File: raiden/connection_manager.py [hunk 5]
```diff
@@ -83,6 +83,8 @@ def __init__(self, raiden, token_network_identifier):
         self.lock = Semaphore()  #: protects self.funds and self.initial_channel_target
         self.api = RaidenAPI(raiden)
 
+        self.channel_sync_state_timeout = 5
```
Comment (LefterisJP) [inline raiden/connection_manager.py]: Can you define this in raiden/constants.py and use it from there?

File: raiden/tests/integration/long_running/flaky/test_token_networks.py [hunk 6]
```diff
@@ -5,11 +5,54 @@
...
 
+def wait_for_reachable_node(sender, receiver):
+    """Wait until a node is reachable"""
+    while True:
+        state = RaidenAPI(sender).get_node_network_state(
+            receiver.address,
+        )
+        if state == NODE_NETWORK_REACHABLE:
+            return
+        gevent.sleep(1)
```
Comment (LefterisJP) [inline raiden/tests/integration/long_running/flaky/test_token_networks.py]: Perhaps put the polling frequency as a `retry_timeout` argument and move this function to the `raiden/waiting.py` module? That's where we have functions that wait like this.

File: raiden/tests/integration/long_running/flaky/test_token_networks.py [hunk 7]
```diff
@@ -5,11 +5,54 @@
...
+    sender_address,
+):
+    """Wait until a first transaction in a channel is received"""
+    while True:
+        receiver_channel = RaidenAPI(receiver).get_channel_list(
+            registry_address=registry_address,
+            token_address=token_address,
+            partner_address=sender_address,
+        )
+        if(
```
Comment (LefterisJP) [inline raiden/tests/integration/long_running/flaky/test_token_networks.py]: nitpick:

It would be a bit nicer to read if you put the condition in a variable and then did the `if` check on the variable. We do that a lot in the code for readability. e.g.:

```python
transaction_received = (
            len(receiver_channel) == 1 and
            receiver_channel[0].partner_state.balance_proof is not None
)
if transaction_received:
    break

File: raiden/tests/integration/long_running/flaky/test_token_networks.py [hunk 8]
```diff
@@ -5,11 +5,54 @@
...
+    while True:
+        state = RaidenAPI(sender).get_node_network_state(
+            receiver.address,
+        )
+        if state == NODE_NETWORK_REACHABLE:
+            return
+        gevent.sleep(1)
+
+
+def wait_for_transaction(
```
Comment (LefterisJP) [inline raiden/tests/integration/long_running/flaky/test_token_networks.py]: same thing as the above function. Check if any of the functions in `waiting.py` fulfill your needs instead. If not please move this also there with a configurable `retry_timeout`

File: raiden/transfer/state.py [hunk 9]
```diff
@@ -1003,6 +1003,13 @@ def __eq__(self, other):
             self.deposit_block_number == other.deposit_block_number
         )
 
+    def __lt__(self, other):
```
Comment (LefterisJP) [inline raiden/transfer/state.py]: Question: The order we impose here does not really matter, right? It just needs to have the `__lt__` operator implemented to solve ordering for `TransactionChannelNewBalance` as per #1796?

File: raiden/connection_manager.py (hunk 2 above)
Comment (pcppcp) [reply raiden/connection_manager.py]: ha, this one was a tricky bug. But there's a possible sync problem:

if nodes A and B are running asynchronously, following can happen
1) A opens a channel
2) B also wants to open a channel at the same time, but A's transaction is already mined -> `self.api.channel_open()` raises `DuplicatedChannelError` (transaction throws)
https://github.com/raiden-network/raiden/blob/c72407fd6c7d1f6e01dc703a1ba6278091038ca2/raiden/connection_manager.py#L298
3) BUT B's state isn't yet synced to a block in which the A's open channel transaction is mined
4) if B proceeds without waiting, and calls `self.api.set_total_channel_deposit()`, an exception will be raised from here
https://github.com/raiden-network/raiden/blob/c72407fd6c7d1f6e01dc703a1ba6278091038ca2/raiden/api/python.py#L290
even if the channel exists on-chain (but not in the raiden state)

File: raiden/transfer/state.py (hunk 9 above)
Comment (pcppcp) [reply raiden/transfer/state.py]: Yes, the order doesn't matter. I am not sure why the heap can't order without the '<' operator implemented - maybe `__eq__()` overrides it?

For any generic class the ordering works (and uses instance address as the key)
```
>>> class X():
...   pass
... 
>>> a = X()
>>> b = X()
>>> a < b
True
>>> b < a
False
```

File: raiden/connection_manager.py (hunk 2 above)
Comment (hackaugusto) [reply raiden/connection_manager.py]: OHHH, I see.

We should move the fix down to the `RaidenAPI` class then. Otherwise things like the rest api will also be susceptible to this race condition.

What about doing all of the following?

- Adding a check in the [`channel_open` method](https://github.com/raiden-network/raiden/blob/c72407fd6c7d1f6e01dc703a1ba6278091038ca2/raiden/api/python.py#L187) to see if the channel already exists in the current state of the node, and if it does `DuplicatedChannelError` is raised
- Adding a [`try-except` around the proxy call](https://github.com/raiden-network/raiden/blob/c72407fd6c7d1f6e01dc703a1ba6278091038ca2/raiden/api/python.py#L222-L225), which handles the `DuplicatedChannelError` exception, calls `waiting.wait_for_newchannel` to make sure the node that lost the race is synchronized, and re-raises the exception

I think this will fix the race condition for the connection manager and the other users.

File: raiden/transfer/state.py (hunk 9 above)
Comment (hackaugusto) [reply raiden/transfer/state.py]: Also, please use [`total_odering`](https://docs.python.org/3/library/functools.html?highlight=total_order#functools.total_ordering)

File: raiden/connection_manager.py (hunk 2 above)
Comment (LefterisJP) [reply raiden/connection_manager.py]: @pcppcp Very nice find! Indeed we got a potential race there.

I second Augusto's suggestion to move the waiting logic/locking in the RaidenAPI so we cover all usage cases and not just this test.

File: raiden/tests/integration/long_running/flaky/test_token_networks.py (hunk 4 above)
Comment (pcppcp) [reply raiden/tests/integration/long_running/flaky/test_token_networks.py]: I have no idea either. I've removed the whole block and it still works....

File: raiden/transfer/state.py (hunk 9 above)
Comment (pcppcp) [reply raiden/transfer/state.py]: nice decorator, thanks for the tip. I wonder why do they require `__eq__` to be implemented if `__lt__` is available.

```
def __eq__(self, o):
  return (not (self < o) ) and (not (o < self))
```

File: raiden/transfer/state.py [hunk 10]
```diff
@@ -996,15 +998,22 @@ def __repr__(self):
         )
 
     def __eq__(self, other):
+        if not isinstance(other, TransactionChannelNewBalance):
+            raise NotImplemented
```
Comment (hackaugusto) [inline raiden/transfer/state.py]: why was this changed? if the `other` is not a `TransactionChannelNewBalance` then the values are not equal

Comment (pcppcp) [reply raiden/transfer/state.py]: apparently it should be `return` instead of `raise`. Will fix that.
